pollux/lib/core/src/main/scala/io/iohk/atala/pollux/core/service/HttpURIDereferencerImpl.scala

```diff
@@ -1,31 +1,38 @@
 package io.iohk.atala.pollux.core.service
 
 import io.iohk.atala.pollux.core.service.URIDereferencerError.{ConnectionError, ResourceNotFound, UnexpectedError}
-import zio.http.*
 import zio.*
+import zio.http.*
+import zio.stream.ZSink
 
 import java.net.URI
+import java.nio.charset.StandardCharsets
 
 class HttpURIDereferencerImpl(client: Client) extends URIDereferencer {
 
   override def dereference(uri: URI): IO[URIDereferencerError, String] = {
-    val result: ZIO[Client, URIDereferencerError, String] = for {
-      response <- Client.request(uri.toString).mapError(t => ConnectionError(t.getMessage))
+    val program = for {
+      url <- ZIO.fromOption(URL.fromURI(uri)).mapError(_ => ConnectionError(s"Invalid URI: $uri"))
+      response <- client
+        .request(Request(url = url))
+        .mapError(t => ConnectionError(t.getMessage))
       body <- response.status match {
         case Status.Ok =>
           response.body.asString.mapError(t => UnexpectedError(t.getMessage))
-        case Status.NotFound if !response.status.isError => ZIO.fail(ResourceNotFound(uri))
-        case _ if response.status.isError =>
-          val err = response match {
-            case Response.GetError(error) => Some(error)
-            case _                        => None
-          }
-          ZIO.fail(UnexpectedError(s"HTTP response error: $err"))
-        case _ =>
-          ZIO.fail(UnexpectedError("Unknown error"))
+        case Status.NotFound =>
+          ZIO.fail(ResourceNotFound(uri))
+        case status if status.isError =>
+          response.body.asStream
+            .take(1024) // Only take the first 1024 bytes from the response body (if any).
+            .runFold(Array.empty[Byte])(_ :+ _)
```

Comment:
This maybe intensive as new array is copied every for next byte. `.runCollect` or mutate original array might be better here. But it's only for 1024 bytes max so it's not that serious :smile:

Reply:
Nice - I'm not the original author but I would be very happy to improve it before it's merged - will ping you if I need help :-)